core/embed/rust/src/micropython/obj.rs

```diff
@@ -134,12 +145,19 @@ impl TryFrom<Obj> for i32 {
 
         // TODO: Avoid type casts on the Python side.
         // SAFETY:
-        //  - Can raise if `obj` is int but cannot fit into `cty::mp_int_t`.
-        if unsafe { ffi::mp_obj_get_int_maybe(obj.as_ptr(), &mut int) } {
-            let int = int.try_into()?;
-            Ok(int)
-        } else {
-            Err(Error::NotInt)
+        //  - `int` is a mutable pointer of the right type.
+        //  - `obj` can be anything uPy understands.
+        // EXCEPTION: Can raise if `obj` is int but cannot fit into `cty::mp_int_t`.
+        let result =
+            catch_exception(|| unsafe { ffi::mp_obj_get_int_maybe(obj.as_ptr(), &mut int) });
+        match result {
+            Ok(true) => Ok(int.try_into()?),
+            Ok(false) => Err(Error::TypeError),
+            Err(_) => {
+                let int64: i64 = obj.try_into()?;
```

Comment:
Why don't we just return an error?

Reply:
Ah, because I got confused and thought that `mp_obj_get_int_maybe` only returns the smallint range. Good catch.

Reply:
fixed in b406bd0a0ff986392c2fccef336a7b9201734023